### PR TITLE
New post about BERT for search applications

### PR DESCRIPTION
**Pull request by thigm85**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by kkraune** [comment]
generally using linebreaks makes it easier to diff / make quick edits - for next time :-)

File: _posts/2020-11-25-fine-tune-bert-basic-transformers-trainer-search-applications.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+There are cases where the inputs to your Transformer model are pairs of sentences, but you want to process each sentence of the pair at different times due to your application’s nature.
+
+![Decorative image](/assets/2020-11-25-fine-tune-bert-basic-transformers-trainer-search-applications/figure_1.jpg)
+<p class="image-credit">Photo by <a href="https://unsplash.com/@alicegrace?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText">Alice Dietrich</a> on <a href="https://unsplash.com/@alicegrace?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText">Unsplash</a></p>
+
+## The search use case
+
+Search applications are one example. They involve a large collection of documents that can be pre-processed and stored before a search action is required. On the other hand, a query triggers a search action, and we can only process it in real-time. Search apps’ goal is to return the most relevant documents to the query as quickly as possible. By applying the tokenizer to the documents as soon as we feed them to the application, we only need to tokenize the query when a search action is required, saving us precious time.
```
Comment (kkraune) [inline _posts/2020-11-25-fine-tune-bert-basic-transformers-trainer-search-applications.md:21]:
```suggestion
Search applications are one example. They involve a large collection of documents that can be pre-processed and stored before a search action is required. On the other hand, a query triggers a search action, and we can only process it in real-time. Search apps’ goal is to return the most relevant documents to the query as quickly as possible. By applying the tokenizer to the documents as soon as we feed them to the application, we only need to tokenize the query when a search action is required, saving time.
```

File: _posts/2020-11-25-fine-tune-bert-basic-transformers-trainer-search-applications.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+from transformers import BertTokenizerFast
+
+model_name = "google/bert_uncased_L-4_H-512_A-8"
+tokenizer = BertTokenizerFast.from_pretrained(model_name)
+
+train_encodings = tokenizer(train_queries, train_docs, truncation=True, padding='max_length', max_length=128)
+val_encodings = tokenizer(val_queries, val_docs, truncation=True, padding='max_length', max_length=128)
+```
+
+In the search case, we create the `create_bert_encodings` function that will apply two different tokenizers, one for the query and the other for the document. In addition to allowing for different query and document `max_length`, we also need to set `add_special_tokens=False` and not use `padding` as those need to be included by our custom code when joining the tokens generated by the `tokenizer`.
```
Comment (kkraune) [inline _posts/2020-11-25-fine-tune-bert-basic-transformers-trainer-search-applications.md:44]:
```suggestion
In the search case, we create the `create_bert_encodings` function that will apply two different tokenizers, one for the query and the other for the document. In addition to allowing for different query and document `max_length`, we also need to set `add_special_tokens=False` and not use `padding`, as those need to be included by our custom code when joining the tokens generated by the `tokenizer`.
```

File: _posts/2020-11-25-fine-tune-bert-basic-transformers-trainer-search-applications.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+![Decorative image](/assets/2020-11-25-fine-tune-bert-basic-transformers-trainer-search-applications/figure_2.jpg)
+<p class="image-credit">Photo by <a href="https://unsplash.com/@steve_j?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText">Steve Johnson</a> on <a href="https://unsplash.com/@steve_j?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText">Unsplash</a></p>
+
+We showed how to customize BERT model encodings to ensure this training and serving compatibility. However, a better approach is to build tools that bridge the gap between training and serving by allowing users to request training data that respects by default the encodings used when serving the model. [pyvespa](https://pyvespa.readthedocs.io/en/latest/index.html) will include such integration to make it easier for [Vespa](https://vespa.ai/) users to train BERT models without having to adjust the encoding generation manually as we did above.
```
Comment (kkraune) [inline _posts/2020-11-25-fine-tune-bert-basic-transformers-trainer-search-applications.md:117]: It is hard to understand the first sentence

**Review by kkraune — APPROVED** [review]
Great post!
